src/sfctl/custom_app.py

```diff
@@ -65,6 +68,167 @@ def validate_app_path(app_path):
+    '''
+    try:
+        cluster_manifest = json.loads(cluster_manifest_content)
+        cluster_namespace = {
+            'ClusterManifest': 'http://schemas.microsoft.com/2011/01/fabric'}
+        cluster_manifest_str = str(cluster_manifest['Manifest'])
+        image_store_connection_string = ET.fromstring(
+            cluster_manifest_str).find(
+                'ClusterManifest:FabricSettings', cluster_namespace).find(
+                    ".//*[@Name='Management']").find(
+                        ".//*[@Name='ImageStoreConnectionString']").attrib['Value'] #pylint: disable=line-too-long
+        return image_store_connection_string
+    except AttributeError:
+        pass
+    except TypeError:
+        return None
+
+def get_imagestore_connection_string(sesh, endpoint): #pylint: disable=invalid-name
+    """
+    Obtains cluster manifest through REST API and parses the
+    image store connection string  from the cluster manifest
+    """
+    try:
+        from urllib.parse import urlparse, urlencode, urlunparse
+    except ImportError:
+        from urllib import urlencode
+        from urlparse import urlparse, urlunparse  # pylint: disable=import-error
+
+    try:
+        url_path = '$/GetClusterManifest'
+        url_parsed = list(urlparse(endpoint))
+        url_parsed[2] = url_path
+        url_parsed[4] = urlencode(
+            {'api-version': '3.0-preview'})
+        url = urlunparse(url_parsed)
+
+        manifest_response = sesh.get(url)
+        return parse_imagestore_connection_string(manifest_response.content)
+    except Exception: #pylint: disable=broad-except
+        return None
+
+def parse_file_share_path(imagestore_connstr):
+    '''
+    Parse the file share path from the image store connection string
+    '''
+    if imagestore_connstr and 'file:' in imagestore_connstr:
+        conn_str_list = imagestore_connstr.split("file:")
+        return conn_str_list[1]
+    else:
+        return False
+
+def get_file_share(sesh, endpoint):
+    """
+    Check to see if fileshare is used as image store connection string in the
+    cluster manifest.
+    """
+    imagestore_connstr = get_imagestore_connection_string(
+        sesh, endpoint)
+    return parse_file_share_path(imagestore_connstr)
+
+def upload_to_fileshare(source, dest, show_progress):
+    """
+    Copies the package from source folder to dest folder.
+    """
+    total_files_count = 0
+    current_files_count = 0
+    for root, _, files in os.walk(source):
+        # Number of uploads is number of files plus number of directories
+        total_files_count += (len(files) + 1)
+
+    for root, _, files in os.walk(source):
+        if not os.path.isdir(root):
+            os.makedirs(root)
+
+        for f in files:
+            rel_path = root.replace(source, '').lstrip(os.sep)
+            dest_path = os.path.join(dest, rel_path)
+            if not os.path.isdir(dest_path):
+                os.makedirs(dest_path)
+
+            shutil.copyfile(
```

Comment:
Can this throw any errors? It may be worthwhile to attempt a retry on transient errors or catch known ones and throw a nice `CLIError`